Mathlib/Algebra/BigOperators/Order.lean

```diff
@@ -816,3 +816,55 @@ theorem IsAbsoluteValue.map_prod [CommSemiring R] [Nontrivial R] [LinearOrderedC
 #align is_absolute_value.map_prod IsAbsoluteValue.map_prod
 
 end AbsoluteValue
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+
+/-- The `positivity` extension which proves that `∑ i in s, f i` is nonnegative if `f` is, and
+positive if each `f i` is and `s` is nonempty.
+
+Note that this does not do any complicated reasoning. In particular, it does not try to feed in the
+`i ∈ s` hypothesis to local assumptions, and the only ways it can prove `s` is nonempty is if there
+is a local `s.Nonempty` hypothesis or `s = (univ : Finset α)` and `Nonempty α` can be synthesized by
+TC inference. -/
+@[positivity Finset.sum _ _]
+def evalFinsetSum : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.sum _ $ι $instα $s $f) =>
+    let i : Q($ι) ← mkFreshExprMVarQ q($ι)
+    have body : Q($α) := Expr.betaRev f #[i]
+    let so : Option Q(Finset.Nonempty $s) ← do
+      try
+        match s with
+        | ~q(@univ _ $fi) => do
+          let _no ← synthInstanceQ q(Nonempty $ι)
+          return some q(Finset.univ_nonempty (α := $ι))
+        | _ => throwError "`s` is not `univ`"
+      catch _ => do
+        let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) | pure none
+        pure (some (.fvar fv))
+    match ← core zα pα body, so with
+    | .nonnegative pb, _ => do
+      let pα' ← synthInstanceQ q(OrderedAddCommMonoid $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars #[i] pb
+      return .nonnegative q(@sum_nonneg $ι $α $pα' $f $s fun i _ ↦ $pr i)
+    | .positive pb, .some (fi : Q(Finset.Nonempty $s)) => do
+      let pα' ← synthInstanceQ q(OrderedCancelAddCommMonoid $α)
+      assertInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars #[i] pb
+      return .positive q(@sum_pos $ι $α $pα' $f $s (fun i _ ↦ $pr i) $fi)
+    | _, _ => pure .none
```

Comment:
What if `body` is positive but `so` is `none`; don't we still want to prove that the sum is non-negative?

Reply:
Fixed. It makes the extension longer but hopefully more robust.